### ISSUE TITLE
Several new metrics for addressing unbalanced classification

### ISSUE THREAD

**Issue by SonghuaHu-UMD** [jdb78/pytorch-forecasting#867]
### Several new metrics for addressing unbalanced classification

I have written several metrics for addressing unbalanced classification, including dice loss, Tversky Loss, weighted cross-entropy, and focal loss. Hope it helps anyone who needed. You can add them before your code and used them in TFT directly.

### Code

```
class CrossEntropy_weight(MultiHorizonMetric):

    def __init__(self, weight: torch.Tensor = None, **kwargs):
        self.weight = weight
        super().__init__()

    def loss(self, y_pred, target):
        loss = F.cross_entropy(y_pred.view(-1, y_pred.size(-1)), target.view(-1), reduction="none",
                               weight=self.weight).view(-1, target.size(-1))
        return loss

    def to_prediction(self, y_pred: torch.Tensor) -> torch.Tensor:
        return y_pred.argmax(dim=-1)

    def to_quantiles(self, y_pred: torch.Tensor, quantiles: List[float] = None) -> torch.Tensor:
        return y_pred


class DiceLoss(MultiHorizonMetric):
    def __init__(self, smooth=1e-5):
        self.smooth = smooth
        super().__init__()

    def loss(self, y_pred, target):
        logits = y_pred.view(-1, y_pred.size(-1))  # N*C, N = B * Dncoder length
        true = target.view(-1)  # N

        # One-hot true
        num_classes = logits.shape[1]
        true_1_hot = torch.eye(num_classes)[true.squeeze()]  # N*C
        probas = F.softmax(logits, dim=1)
        true_1_hot = true_1_hot.type(logits.type())

        # Calculate dice
        intersection = torch.sum(probas * true_1_hot, dim=0)  # C
        cardinality = torch.sum(probas + true_1_hot, dim=0)  # C
        dice_loss = ((2. * intersection + self.smooth) / (cardinality + self.smooth)).mean()

        return (1 - dice_loss) * true_1_hot.size(0)

    def to_prediction(self, y_pred: torch.Tensor) -> torch.Tensor:
        return y_pred.argmax(dim=-1)

    def to_quantiles(self, y_pred: torch.Tensor, quantiles: List[float] = None) -> torch.Tensor:
        return y_pred


class DiceLoss_weight(MultiHorizonMetric):
    def __init__(self, smooth=1e-5):
        self.smooth = smooth
        super().__init__()

    def loss(self, y_pred, target):
        logits = y_pred.view(-1, y_pred.size(-1))  # N*C, N = B * Encoder length
        true = target.view(-1)  # N

        # One-hot true
        num_classes = logits.shape[1]
        true_1_hot = torch.eye(num_classes)[true.squeeze()]  # N*C
        probas = F.softmax(logits, dim=1)
        true_1_hot = true_1_hot.type(logits.type())

        # Calculate weight: 1/number**2
        ground_o = torch.sum(true_1_hot, dim=0).float()  # C
        w = 1. / (ground_o * ground_o).clamp(min=self.smooth)

        intersection = torch.sum(probas * true_1_hot, dim=0)  # C
        cardinality = torch.sum(probas + true_1_hot, dim=0)  # C
        dice_loss = ((2. * intersection * w + self.smooth) / (cardinality * w + self.smooth)).mean()

        return (1 - dice_loss) * true_1_hot.size(0)

    def to_prediction(self, y_pred: torch.Tensor) -> torch.Tensor:
        return y_pred.argmax(dim=-1)

    def to_quantiles(self, y_pred: torch.Tensor, quantiles: List[float] = None) -> torch.Tensor:
        return y_pred


class TverskyLoss(MultiHorizonMetric):
    def __init__(self, smooth=1e-5, alpha=0.1, beta=0.9):
        self.smooth = smooth
        self.alpha = alpha
        self.beta = beta
        super().__init__()

    def loss(self, y_pred, target):
        logits = y_pred.view(-1, y_pred.size(-1))  # N*C, N = B * Encoder length
        true = target.view(-1)  # N

        # One-hot true
        num_classes = logits.shape[1]
        true_1_hot = torch.eye(num_classes)[true.squeeze()]  # N*C
        probas = F.softmax(logits, dim=1)
        true_1_hot = true_1_hot.type(logits.type())

        intersection = torch.sum(probas * true_1_hot, dim=0)  # C
        fps = torch.sum(probas * (1. - true_1_hot), dim=0)
        fns = torch.sum((1. - probas) * true_1_hot, dim=0)
        cardinality = intersection + self.alpha * fps + self.beta * fns

        dice_loss = ((intersection + self.smooth) / (cardinality + self.smooth)).mean()

        return (1 - dice_loss) * true_1_hot.size(0)

    def to_prediction(self, y_pred: torch.Tensor) -> torch.Tensor:
        return y_pred.argmax(dim=-1)

    def to_quantiles(self, y_pred: torch.Tensor, quantiles: List[float] = None) -> torch.Tensor:
        return y_pred


class FocalLoss(MultiHorizonMetric):

    def __init__(self, weight: torch.Tensor = None, gamma=2., reduction='none', **kwargs):
        self.weight = weight
        self.gamma = gamma
        self.reduction = reduction
        super().__init__()

    def loss(self, input_tensor, target_tensor):
        log_prob = F.log_softmax(input_tensor, dim=-1)
        prob = torch.exp(log_prob)
        nll_input = ((1 - prob) ** self.gamma) * log_prob
        loss = F.nll_loss(nll_input.view(-1, input_tensor.size(-1)),
                          target_tensor.view(-1), weight=self.weight, reduction='none').view(-1, target_tensor.size(-1))
        return loss

    def to_prediction(self, y_pred: torch.Tensor) -> torch.Tensor:
        return y_pred.argmax(dim=-1)

    def to_quantiles(self, y_pred: torch.Tensor, quantiles: List[float] = None) -> torch.Tensor:
        return y_pred
```

**Comment by georgeblck** [comment]
I wrote it like this

```
class MaskedMAE(MultiHorizonMetric):
    """
    Mean average absolute error.

    Defined as ``(y_pred - target).abs()``
    """
    def __init__(self, weight: torch.Tensor = None, **kwargs):
        self.weight = weight
        super().__init__()

    def loss(self, y_pred, target):
        loss = (self.to_prediction(y_pred) - target).abs()*self.weight
        return loss
```

Calling it like this
```
MaskedMAE(torch.cat((torch.zeros(setting.max_prediction_length-96,device=torch.device("cuda")),torch.ones(96,device=torch.device("cuda"))))),  # metric to optimize
```
